### ISSUE TITLE
Support a way for the user to provide the output TextWriter(s)

### ISSUE THREAD

**Issue by alexrp** [serilog/serilog-sinks-console#102]
Hi,

My [`System.Terminal` project](https://github.com/alexrp/system-terminal) is a complete replacement for `System.Console` centered around VT100(-ish) terminals on Windows/Unix. One of the caveats of the project is that you can't use `System.Console` at all (due to some implementation details in CoreCLR) so this sink breaks.

`System.Terminal` exposes `Stream` properties for the standard I/O streams which can easily be wrapped in e.g. a `StreamWriter`, but it doesn't seem like there's any way to tell this sink to use them. Would it be reasonable to allow the user to pass in `TextWriter` objects for stdout/stderr when configuring the sink?

**Comment by nblumhardt** [comment]
Hi @alexrp - thanks for the heads-up!

Working without touching System.Console seems like a non-goal for this project; I don't think it's something we'd pursue at this stage.

Perhaps forking it to create a "Terminal" sink and bundling it/shipping it alongside your project would be a decent path forwards?

Cheers!
Nick

**Comment by alexrp** [comment]
I did consider forking and just basically `s/Console/Terminal/g`, and also tried writing my own sink (which I got about halfway with until reconsidering). The issue is that there's actually quite a lot of code in this sink. Either of these options would mean I'd have to actively keep in sync with upstream or make it a 'hard' fork where I just maintain it on my own. These aren't particularly appealing options.

That said, I get that "don't use `System.Console`" is a bit of an unusual and perhaps untenable ask for a project like this.

Here's something that could be considered:

1. Use [this Roslyn analyzer](https://github.com/dotnet/roslyn-analyzers#microsoftcodeanalysisbannedapianalyzers) to ban `System.Console` usage in this sink. (Similar to what I do in `System.Terminal`.)
2. Make all `System.Console` usage go through an `IConsole` interface of sorts. There's precedent for this in other projects, e.g. here: [1](https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/IConsole.cs), [2](https://github.com/dotnet/command-line-api/tree/main/src/System.CommandLine/IO)
3. Provide a `SystemConsole` implementation of `IConsole` that this sink will use by default. (Suppress the `RS0030` banned API diagnostic in this class.)
4. Allow the user to optionally pass in an `IConsole` implementation when configuring this sink.

This would then allow me to provide a minimal NuGet package that just contains an `IConsole` implementation and maybe a few extension methods to easily register it with this sink. The analyzer would ensure that no accidental `System.Console` usage occurs in this sink going forward (without deliberately adding the functionality to `IConsole` first).

I could probably cook up a PR for this if you think this wouldn't cause too much friction for developers of this sink. OTOH, I totally understand if you'd rather not do this.
